File: src/org/labrad/qubits/Device.java
package org.labrad.qubits;

import java.util.List;
import java.util.Map;

import org.labrad.qubits.channels.Channel;

import com.google.common.base.Preconditions;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;

public class Device {
	private String name;
	private List<Channel> channels = Lists.newArrayList();
	private Map<String, Channel> channelsByName = Maps.newHashMap();
	
	public Device(String name) {
		this.name = name;
	}
	
	public String getName() {
		return name;
	}
	
	public void addChannel(Channel ch) {
		channels.add(ch);
		channelsByName.put(ch.getName(), ch);
	}
	
	public List<Channel> getChannels() {
		return Lists.newArrayList(channels);
	}
	
	@SuppressWarnings("unchecked")
	public <T extends Channel> List<T> getChannels(Class<T> cls) {
		List<T> matches = Lists.newArrayList();
		for (Channel chan : channels) {
			if (cls.isInstance(chan)) {
				matches.add((T)chan);
			}
		}
		return matches;
	}
	
	public Channel getChannel(String name) {
		Preconditions.checkArgument(channelsByName.containsKey(name),
			"Device '%s' has no channel named '%s'", getName(), name);
		return channelsByName.get(name);
	}
	
	@SuppressWarnings("unchecked")
	public <T extends Channel> T getChannel(String name, Class<T> cls) {
		Preconditions.checkArgument(channelsByName.containsKey(name),
			"Device '%s' has no channel named '%s'", getName(), name);
		Channel ch = channelsByName.get(name);
		Preconditions.checkState(cls.isInstance(ch),
			"Channel '%s' is not of type '%s'", name, cls.getName());
		return (T)ch;
	}
	
	public <T extends Channel> T getChannel(Class<T> cls) {
		List<T> channels = getChannels(cls);
		Preconditions.checkState(channels.size() == 1,
			"Device '%s' has more than one channel of type '%s'", getName(), cls.getName());
		return channels.get(0);
	}
}

File: QubitServer/src/main/scala/org/labrad/qubits/Device.scala
package org.labrad.qubits

import org.labrad.qubits.channels.Channel
import scala.collection.mutable
import scala.reflect.ClassTag

/**
 * Model of a device, essentially a collection of named channels of various types.
 * Each channel connects various physical devices (FastBias, Preamp, GHzDac) to some
 * controllable parameter on the device (flux bias, microwave, etc.).
 *
 * @author maffoo
 */
class Device(val name: String, val channels: Seq[Channel]) {

  private val channelsByName = channels.map(ch => ch.name -> ch).toMap

  /**
   * Get all defined channels.
   * @return
   */
  def getChannels(): Seq[Channel] = {
    channels
  }

  /**
   * Get all channels of a particular type.
   * @param <T>
   * @param cls
   * @return
   */
  def getChannels[T <: Channel : ClassTag]: Seq[T] = {
    channels.collect { case ch: T => ch }
  }

  /**
   * Get a channel by name (of any type).
   * @param name
   * @return
   */
  def getAnyChannel(name: String): Channel = {
    channelsByName.getOrElse(name, sys.error(s"Device ${this.name} has no channel named $name"))
  }

  /**
   * Get a channel by name and type.
   * @param <T>
   * @param name
   * @param cls
   * @return
   */
  def getChannel[T <: Channel](name: String)(implicit tag: ClassTag[T]): T = {
    getAnyChannel(name) match {
      case ch: T => ch
      case _ => sys.error(s"Channel $name is not of type $tag")
    }
  }

  /**
   * Get a channel by type only.  This fails if there is more than one channel with the desired type.
   * @param <T>
   * @param cls
   * @return
   */
  def getChannel[T <: Channel](implicit tag: ClassTag[T]): T = {
    getChannels[T] match {
      case Seq(ch) => ch
      case Seq() => sys.error(s"Device $name has no channel of type $tag")
      case _ => sys.error(s"Device $name has more than one channel of type $tag")
    }
  }
}
